src/org/zaproxy/zap/spider/URLCanonicalizer.java

```diff
@@ -446,4 +450,83 @@ private static String normalizePath(final String path) {
 		return path.replace("%7E", "~").replace(" ", "%20");
 	}
 
+	/**
+	 * A query parameter, with non-{@code null} name and value.
+	 * <p>
+	 * The query parameters are ordered by name and value.
+	 */
+	private static class QueryParameter implements Comparable<QueryParameter> {
+
+		private final String name;
+		private final String value;
+
+		public QueryParameter(String name, String value) {
+			if (name == null) {
+				throw new IllegalArgumentException("Parameter name must not be null.");
+			}
+			if (value == null) {
+				throw new IllegalArgumentException("Parameter value must not be null.");
+			}
+			this.name = name;
+			this.value = value;
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public String getValue() {
+			return value;
+		}
+
+		@Override
+		public int compareTo(QueryParameter other) {
+			if (other == null) {
+				return 1;
+			}
+			int result = name.compareTo(other.name);
+			if (result != 0) {
+				return result;
+			}
+			return value.compareTo(other.value);
+		}
+
+		@Override
+		public int hashCode() {
```

Comment:
hashCode()....it was over here: https://github.com/thc202/zaproxy/commit/f2fb322ad21f0a5072accb0663426df18c2106c3#diff-24025444157a86c23540fcbb1c660144R106